contracts/zero-ex/tests/Multiplex.t.sol

```diff
@@ -0,0 +1,1453 @@
+        entries = new LogDescription[](5);
+        entries[0] = first;
+        entries[1] = second;
+        entries[2] = third;
+        entries[3] = fourth;
+        entries[4] = fifth;
+    }
+
+    function makeArray(
+        LogDescription memory first,
+        LogDescription memory second,
+        LogDescription memory third,
+        LogDescription memory fourth,
+        LogDescription memory fifth,
+        LogDescription memory sixth
+    ) private pure returns (LogDescription[] memory entries) {
+        entries = new LogDescription[](6);
+        entries[0] = first;
+        entries[1] = second;
+        entries[2] = third;
+        entries[3] = fourth;
+        entries[4] = fifth;
+        entries[5] = sixth;
+    }
+
+    function makeArray(bytes32 first) private pure returns (bytes32[] memory data) {
+        data = new bytes32[](1);
+        data[0] = first;
+    }
+
+    function mintTo(address token, address recipient, uint256 amount) private {
+        if (token == address(weth)) {
+            IEtherTokenV06(token).deposit{value: amount}();
+            WETH9V06(payable(token)).transfer(recipient, amount);
+        } else {
+            TestMintableERC20Token(token).mint(recipient, amount);
+        }
+    }
+
+    function createUniswapV2Pool(
+        TestUniswapV2Factory factory,
+        IERC20TokenV06 tokenA,
+        IERC20TokenV06 tokenB,
+        uint112 balanceA,
+        uint112 balanceB
+    ) private returns (TestUniswapV2Pool pool) {
+        pool = factory.createPool(tokenA, tokenB);
+        mintTo(address(tokenA), address(pool), balanceA);
+        mintTo(address(tokenB), address(pool), balanceB);
+
+        (uint112 balance0, uint112 balance1) = tokenA < tokenB ? (balanceA, balanceB) : (balanceB, balanceA);
+        pool.setReserves(balance0, balance1, 0);
+    }
+
+    function createUniswapV3Pool(
+        TestUniswapV3Factory factory,
+        IERC20TokenV06 tokenA,
+        IERC20TokenV06 tokenB,
+        uint112 balanceA,
+        uint112 balanceB
+    ) private returns (TestUniswapV3Pool pool) {
+        pool = factory.createPool(tokenA, tokenB, POOL_FEE);
+        mintTo(address(tokenA), address(pool), balanceA);
+        mintTo(address(tokenB), address(pool), balanceB);
+    }
+
+    function encodeFractionalFillAmount(uint256 frac) private pure returns (uint256) {
+        return HIGH_BIT + (frac * 1e16);
+    }
+
+    function describeLog(
+        bytes32[] memory topics,
+        bytes memory data,
+        bytes memory dataMask
+    ) private pure returns (LogDescription memory) {
+        assert(dataMask.length == 0 || dataMask.length == data.length);
+        return LogDescription(topics, data, dataMask);
+    }
+
+    function describeLog(bytes32[] memory topics, bytes memory data) private pure returns (LogDescription memory) {
+        return describeLog(topics, data, hex"");
+    }
+
+    function assertLogs(LogDescription[] memory expected) private {
```

Comment:
What's the purpose of this function over [`vm.expectEmit`](https://book.getfoundry.sh/cheatcodes/expect-emit)?

Reply:
This function can mask out parts of the log's data when comparing the expected logs with the emitted logs. We can do a sort of "fuzzy match" on logs with this. In these tests it is used mostly for masking out the `value` field of `Transfer` events, but it is also used to mask out several fields from a `MintTransform` event. I have yet to encounter a case in the original typescript tests where we would need to mask topics, but if we ever do encounter that case then we would add that ability to this function as well.

We will want to mask pieces of the logs out whenever the component we are testing interacts with something outside itself and we want to know some details of their interaction but do not care about all of the details. Some of those details we ignore can even be incorrect if something is wrong with the "other" component, but as long as it doesn't affect what we are testing then it is good that we can ignore it. I imagine that this would be especially useful for something like metatransactions, but I haven't looked at those tests to confirm.

Reply:
Afaik `vm.expectEmit` cannot match _parts_ of the data, only ignore it or match it entirely.

Reply:
Ahh, you're right. Is that feature really this important that we're willing to add this much complexity to our testing infra? It makes me itchy